python/pyarrow/types.pxi

```diff
@@ -420,7 +420,7 @@ cdef class Field:
         return self.__str__()
 
     def __hash__(self):
-        return hash((self.field.name(), self.type.id))
+        return hash((self.field.name(), self.type, self.field.nullable()))
```

Comment:
I guess We should hash the metadata too. 
The hashing could be factored out, it's equivalent with

```py
hash(self.__reduce__()[1])
```

Reply:
IMO the current form is preferred. The `__reduce__` variant is cryptic.
And no, I don't think we should hash the metadata. A hash function should be fast while minimizing _real-world_ collisions.

Reply:
Great explanation, thanks!